### PR TITLE
Si matching

### PR DESCRIPTION
**Pull request by phlptp**
add unit domains in an effort to better match SI standards for some specific units.

Add additional documentation

Address Issue #172

### CODE REVIEW

File: test/test_unit_strings.cpp [hunk 1]
```diff
@@ -101,6 +101,56 @@ TEST(unitStrings, prefixes)
...
+    EXPECT_EQ(to_string(unit_from_string("nm")), "nm");
+    EXPECT_EQ(to_string(unit_from_string("pm")), "pm");
+    EXPECT_EQ(to_string(unit_from_string("fm")), "fm");
+    EXPECT_EQ(to_string(unit_from_string("am")), "am");
+    EXPECT_EQ(to_string(unit_from_string("zm")), "zm");
+    EXPECT_EQ(to_string(unit_from_string("ym")), "ym");
+    EXPECT_EQ(to_string(unit_from_string("dm")), "dm");
+
+    EXPECT_EQ(to_string(unit_from_string("km")), "km");
+    EXPECT_EQ(to_string(unit_from_string("Km")), "km");
```
Comment (SimonHeybrock) [inline test/test_unit_strings.cpp:118]: Any particular reason for supporting uppercase `K` here?

File: units/CMakeLists.txt [hunk 2]
```diff
@@ -46,12 +69,14 @@ elseif(UNITS_BUILD_OBJECT_LIBRARY)
 
     if(UNITS_NAMESPACE)
         target_compile_definitions(units PUBLIC -DUNITS_NAMESPACE=${UNITS_NAMESPACE})
-
-        if(UNITS_BASE_TYPE)
-            target_compile_definitions(
-                units PUBLIC -DUNITS_BASE_TYPE=${UNITS_BASE_TYPE}
-            )
-        endif()
```
Comment (SimonHeybrock) [inline units/CMakeLists.txt:54]: Unreleated, but here is a useful tool, https://github.com/cheshirekow/cmake_format, basically clang-format but for cmake.

File: units/units.hpp [hunk 3]
```diff
@@ -1788,6 +1788,20 @@ inline fixed_precise_measurement sqrt(const fixed_precise_measurement& meas)
...
+UNITS_EXPORT int setUnitsDomain(int newDomain);
+
+/** specify a domain to use in unit translation for some ambiguous units*/
+namespace domains {
+    constexpr std::uint32_t defaultDomain{0U};
+    constexpr std::uint32_t ucum{1U};
+    constexpr std::uint32_t cooking{2U};
+    constexpr std::uint32_t astronomy{4U};
+    constexpr std::uint32_t nuclear{8U};
+    constexpr std::uint32_t surveying{9U};
```
Comment (SimonHeybrock) [inline units/units.hpp:1800]: So surveying is `ucum` and `nuclear`?

File: units/units.cpp [hunk 4]
```diff
@@ -144,12 +144,14 @@ static const umap base_unit_names{
                                        // 1000 m^3 not mega meters cubed
     {kg, "kg"},
     {mol, "mol"},
+    {unit(0.1, m), "dm"},  // don't want to use deci in most conversion contexts
```
Comment (SimonHeybrock) [inline units/units.cpp:147]: I believe `dL` or `dl` are also quite common? Also, I don't know if you support decibel, but if you do, there is `dB`.

File: units/units.cpp [hunk 5]
```diff
@@ -144,12 +144,14 @@ static const umap base_unit_names{
...
+                           // but for meter it is fine
     {A, "A"},
     {A * h, "Ah"},
     {V, "V"},
     {s, "s"},
     // this is so Gs doesn't get used which can cause issues
-    {giga * s, "Bs"},
+    {giga * s, "Bs"},  // B is for billion
```
Comment (SimonHeybrock) [inline units/units.cpp:154]: Is this comment (regarding `Gs`) still true with domains?

File: units/units.cpp [hunk 6]
```diff
@@ -1710,6 +1738,55 @@ static double getPrefixMultiplier(char p)
...
 }
 
+/// Generate the prefix multiplier for SI units
+static double getStrictSIPrefixMultiplier(char p)
+{
+    switch (p) {
+        case 'm':
+            return 0.001;
+        case 'k':
+        case 'K':
```
Comment (SimonHeybrock) [inline units/units.cpp:1748]: Is this actually SI?

File: units/units.cpp [hunk 7]
```diff
@@ -1710,6 +1738,55 @@ static double getPrefixMultiplier(char p)
...
+            return 1e6;
+        case 'u':
+        case '\xB5':  // latin-1 encoding "micro"
+            return 1e-6;
+        case 'd':
+            return 0.1;
+        case 'c':
+            return 0.01;
+        case 'h':
+        case 'H':
```
Comment (SimonHeybrock) [inline units/units.cpp:1760]: Same question as for kilo, is uppercase actually SI?

File: units/units.cpp [hunk 8]
```diff
@@ -3673,7 +3783,6 @@ static const smap base_unit_vals{
     {"mps", precise::m / precise::s},
     {"eV", precise::energy::eV},
     {"bev", precise_unit(1e9, precise::energy::eV)},
-    {"EV", precise::energy::eV},
```
Comment (SimonHeybrock) [inline units/units.cpp:3676]: 👍

**Review by SimonHeybrock — COMMENTED** [review]
Looks very promising. 👍

File: docs/user-guide/Unit_domains.rst [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+.. code-block:: c++
+
+   setUnitsDomain(code);
+
+
+with the code using one of those found in the `units::domains` namespace.
+this domain will be used unless another is specified through the match flags.
+
+The default domain can be set at compile time through the `UNITS_DEFAULT_DOMAIN` definition which is available through CMake options or specified before compilation.
```
[the remark this answers is not in the document]
Comment (phlptp) [reply docs/user-guide/Unit_domains.rst:95]: UNITS_DOMAIN in CMake gets translated as a string to specify UNITS_DEFAULT_DOMAIN in the C++ code.     I should make this clearer.

File: test/test_unit_strings.cpp (hunk 1 above)
Comment (phlptp) [reply test/test_unit_strings.cpp:118]: Correct or not I have seen `Kg` get used in a number of contexts for kilogram.  I suppose I could just remove the support for general upper case 'K' here and specifically support `Kg`

File: units/CMakeLists.txt (hunk 2 above)
Comment (phlptp) [reply units/CMakeLists.txt:54]: I think that gets run as part of the automatic formatting sequence so I have used it.

File: units/units.hpp (hunk 3 above)
Comment (phlptp) [reply units/units.hpp:1800]: guess I need to double check the bit sequences again

File: units/units.cpp (hunk 5 above)
Comment (phlptp) [reply units/units.cpp:154]: `Gs` is a symbol for Gauss so the domains haven't affected that

File: units/units.cpp (hunk 4 above)
Comment (phlptp) [reply units/units.cpp:147]: I will add a `dl` exception.  dB is handled differently since that is an equation unit

File: units/units.cpp (hunk 7 above)
Comment (phlptp) [reply units/units.cpp]: I removed support for 'K' and 'H' in the strict SI conversions

File: units/units.cpp (hunk 6 above)
Comment (phlptp) [reply units/units.cpp]: removed the capital K from the strict SI for kilo.

**Pull request merged**